main/src/library/RedBlackTree.flix

```diff
@@ -679,53 +679,76 @@ namespace RedBlackTree {
+    /// the threads.
+    ///
+    @Parallel
+    pub def parMaximumBy(n: Int32, cmp: (k, v, k, v) -> Comparison, t: RedBlackTree[k, v]): Option[(k, v)] & Impure =
+        parLimitBy(n, GreaterThan, cmp, t)
+
+    ///
+    /// Helper function for `minimumBy` & `maximumBy`.
+    ///
+    /// Applies `cmp` over the tree `t` in parallel and optionally returns the min/max (decided by `decider`)
+    /// element according to `cmp`.
+    ///
+    /// The implementation spawns `n` threads each applying `cmp` sequentially
+    /// from left to right on some subtree that is disjoint from the rest of
+    /// the threads.
+    ///
+    @Parallel
+    def parLimitBy(n: Int32, decider: Comparison, cmp: (k, v, k, v) -> Comparison, t: RedBlackTree[k, v]): Option[(k, v)] & Impure =
         if (n <= 0)
-            parSeqMinimumBy(cmp, t)
+            parSeqLimitBy(decider, cmp, t)
         else
             match t {
                 case Leaf                => None
                 case DoubleBlackLeaf     => None
                 case Node(_, a, k, v, b) =>
                     let chanL = chan Option[(k, v)] 1;
                     let chanR = chan Option[(k, v)] 1;
-                    spawn (chanL <- parMinimumBy((n - 2) / 2, cmp, a));
-                    spawn (chanR <- parMinimumBy((n - 2) / 2, cmp, b));
-                    def smallest(kl, vl, kr, vr) = if (cmp(kl, vl, kr, vr) == LessThan) (kl, vl) else (kr, vr);
+                    spawn (chanL <- parLimitBy((n - 2) / 2, decider, cmp, a));
+                    spawn (chanR <- parLimitBy((n - 2) / 2, decider, cmp, b));
+                    def selector(kl, vl, kr, vr) = if (cmp(kl, vl, kr, vr) == decider) (kl, vl) else (kr, vr);
                     let a1 = <- chanL;
                     let res = match a1 {
                         case None         => (k, v)
-                        case Some(kl, vl) => smallest(kl, vl, k, v)
+                        case Some(kl, vl) => selector(kl, vl, k, v)
                     };
                     let b1 = <- chanR;
                     match b1 {
                         case None         => Some(res)
                         case Some(kr, vr) =>
                             let (ks, vs) = res;
-                            Some(smallest(ks, vs, kr, vr))
+                            Some(selector(ks, vs, kr, vr))
                     }
             }
 
     ///
     /// Applies `cmp` over the tree `t` sequentially and optionally returns the lowest
     /// element according to `cmp`.
     ///
-    /// Note that this is equivalent to `seqMinimumBy` but for performance reasons -- to avoid
-    /// megamorphic calls -- we use a copy here.
-    ///
-    def parSeqMinimumBy(cmp: (k, v, k, v) -> Comparison & ef, t: RedBlackTree[k, v]): Option[(k, v)] & ef = match t {
+    def parSeqLimitBy(decider: Comparison, cmp: (k, v, k, v) -> Comparison & ef, t: RedBlackTree[k, v]): Option[(k, v)] & ef = match t {
```

Comment:
I think the role of decider is to know if you are looking for the smallest/largest element. And then `cmp` is used for the comparison. My question is: Could we not have a function like `(k, v, k, v) -> (k, v)` that simply returns the right element (i.e. either the smallest or largest)?

Reply:
Something like `...f(cmp(kl, vl, kr, vr))...`, `f: Comparison -> (k, v)`?

Reply:
I think I understand what you mean. I'll push a draft in a bit

Reply:
Hopefully it is a simplification across the board :)